lib/filesystem.nix

```diff
@@ -61,6 +61,16 @@ in
     in go (if isDir then file else parent);
 
 
+  /*
+    Given a directory and a regex, return a list of basenames of
+    files in that directory which match the regex.
+  */
+  listFilesMatching = regex: dir:
+    lib.lists.filter (x: x!=null)
+      (lib.mapAttrsToList
+        (k: _: if builtins.match regex k == null then null else k)
+        (builtins.readDir dir));
```

Comment:
I'm not sure if this is a good function to introduce into `lib`. I think it would be cleaner to have a `pkgs/tools/package-management/nix/versions` directory where each entry corresponds to a version, then you don't need a regex filter and can just use `readDir`. Same for other packages if necessary.

Reply:
> I think it would be cleaner to have a `pkgs/tools/package-management/nix/versions` directory where each entry corresponds to a version, then you don't need a regex filter and can just use `readDir`

I agree.  Done: e618baa4be9ea0fe1d2124300b6f42a54999b89c.

The original naming scheme came from [this request](https://github.com/NixOS/nixpkgs/pull/211172#discussion_r1083375585).

I will update all of this to match RFC140 when it is finished and implemented.  However between now and then, having `pkgs/tools/package-management/nix/` package split up into file-per-version will save rebasing pain on every commit that touches this directory.

I assume RFC140's implementation will include something like (and more general than) the three `version_to_` helper functions, so those will go away.